### PR TITLE
Detecting popstate to solve probem with browser:forward

### LINKED ISSUE #3642: Browser History partially broken, back working but forward not
## Current behavior
If going from some dynamic link to another, like click through categories, or going from category to a product and using the browser:back functionality everything is fine. But after using browser.back there is just a short "flicker" of the browser:forward button and afterwards it is not usable. Browser:back keeps working.
Originating from static links, like homepage, going to a product > browser:back results in a working browser:forward mechanic.

## Expected behavior
After using browser:back every page should be able to have a sane browser history and be able to use browser:forward.

## Steps to reproduce the issue
From home, use a category from the menu, click on a product, use browser:back and browser:forward is not working.
Coming from home, going directly to a product > browser:back then browser:forward is working.

## Can you handle fixing this bug by yourself?

- [ ] YES
- [ x] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [x ] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ x] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details
- Browser: chrome
- OS: linux

### PR DESCRIPTION
**Pull request by andrzejewsky**
### Related issues
closes #3642 

### Short description and why it's useful
It solves the problem with `history:forward`.

## Description
This is really hacky workaround to solve that problem with `history:forward` action. I aded `popstate` listener to detect whether user has used back or forward button and if that happens, I'm calling replace instead of push which won't break the history stack. We can call just `next()`, it also works with that, but it breaks some pages (probably because some parameters are missing). 

I think we should use that kind of hooks only for specific things and use arguments to the `next` call only when we really need it. For sure not for coordinating most of routing. I really know that it was implemented in this way for specific reason, but maybe one day we'll try to find out how to avoid something like that?

I'm not really proud of that solution but it works and solves the problem. Maybe we won't even merge it, please have a look @pkarw @filrak @patzick.

### Which environment this relates to
- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [x] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

### Contribution and currently important rules acceptance
- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/lib/router-manager.ts [hunk 1]
```diff
@@ -88,6 +89,26 @@ const RouterManager = {
...
     });
+  },
+  watchHistory: function () {
+    if (typeof window === 'undefined') {
+      return
+    }
+
+    window.popstateDetected = false
+    window.addEventListener('popstate', () => {
+      console.log('POPSTATE')
```
Comment (pkarw) [inline core/lib/router-manager.ts]: please remove this console log

**Comment by gibkigonzo** [comment]
There is still some other issue with history:
1. go to https://test.storefrontcloud.io
2. go to some product
3. go to some category
4. refresh page
5. click back button and history will crash

**Comment by gibkigonzo** [comment]
I think that we can fix issue that I described above :point_up: by adding something like this:
```
        if (dynamicRoute) {
          const isPageRefreshed = dynamicRoute.fullPath === from.fullPath
          next({ ...dynamicRoute, replace: RouterManager.isPopState() || isPageRefreshed })
        } else {
```

File: core/lib/router-manager.ts [hunk 2]
```diff
@@ -88,6 +89,26 @@ const RouterManager = {
     this._callbacks.forEach(callback => {
       callback()
     });
+  },
+  watchHistory: function () {
+    if (typeof window === 'undefined') {
```
Comment (patzick) [inline core/lib/router-manager.ts]: we have `isServer` helper for this

**Comment by andrzejewsky** [comment]
@patzick Really thanks, I knew it there is a better way to do this! I totally forgot about about observables, I'll fix it soon.

**Comment by andrzejewsky** [comment]
@patzick updated

**Comment by filrak** [comment]
Just from code organization pov shouldn't this helper be a part of url dispatcher module? Is it still useful without it?

**Comment by alinadivante** [comment]
ok, after testing, it looks fine for me

**Pull request merged**